src/main/java/echoserver/ServerSocketWrapper.java

```diff
@@ -0,0 +1,66 @@
+package echoserver;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.PrintWriter;
+import java.net.ServerSocket;
+import java.net.Socket;
+
+public class ServerSocketWrapper extends Thread implements SocketWrapper {
+
+    private final int portNumber;
+    private ServerSocket serverSocket;
+    private Socket socket;
+    private BufferedReader input;
+    private PrintWriter output;
+    private String clientMessage = "";
+
+    public ServerSocketWrapper(int portNumber) {
+        this.portNumber = portNumber;
+    }
+
+    @Override
+    public void run() {
```

Comment:
I'd avoid using threads initially. It would probably be a good idea to keep your code for dealing with threads separate from the code dealing with echoing the messages.

I've not tried, but I don't think this will do what you think it should currently. Have a look at the behaviour of `accept()`. Why do you think the current implementation (including how `ServerSocketWrapper` is called from `EchoServer`) won't handle concurrent requests?

Reply:
Thanks, Ashley. I'll remove this for now. Currently, concurrent requests won't be handled, because `.accept()` is only being called once, and therefore will only listen once for a client and accept it